### PR TITLE
DeviceNodeGroup

### PR DESCRIPTION
**Pull request by lnotspotl**
This PR introduces a couple of new items into `depthai`

1. `DeviceNodeGroup` class that works just like `NodeGroup` with the exception that it sets `device` for all subnodes that inherit from `DeviceNode`
2. `DetectionNetwork` now inheritcs from the newly introduced `DeviceNodeGroup` class
3. For convenience reasons, `DetectionNetwork`, `NeuralNetwork` and `DetectionParser` can now be set up via the `setFromModelzoo` method. This method accepts a model description, downloads a model from modelzoo and sets it. The device type (RVC2, RVC3, RVC4) is set automatically based on the device.

Clickup tasks: 
https://app.clickup.com/t/86bzrkrp1
https://app.clickup.com/t/86bzrkx44

### CODE REVIEW

File: bindings/python/src/pipeline/node/DetectionNetworkBindings.cpp [hunk 1]
```diff
@@ -76,6 +77,8 @@ void bind_detectionnetwork(pybind11::module& m, void* pCallstack) {
         .def("getNumInferenceThreads", &DetectionNetwork::getNumInferenceThreads, DOC(dai, node, DetectionNetwork, getNumInferenceThreads))
         .def("setNNArchive", py::overload_cast<const NNArchive&>(&DetectionNetwork::setNNArchive), py::arg("archive"), DOC(dai, node, DetectionNetwork, setNNArchive))
         .def("setNNArchive", py::overload_cast<const NNArchive&, int>(&DetectionNetwork::setNNArchive), py::arg("archive"), py::arg("numShaves"), DOC(dai, node, DetectionNetwork, setNNArchive))
+        .def("setFromModelzoo", py::overload_cast<NNModelDescription, bool>(&DetectionNetwork::setFromModelzoo), py::arg("description"), py::arg("useCached"), DOC(dai, node, DetectionNetwork, setFromModelzoo))
+        .def("setFromModelzoo", py::overload_cast<NNModelDescription, int, bool>(&DetectionNetwork::setFromModelzoo), py::arg("description"), py::arg("numShaves"), py::arg("useCached"), DOC(dai, node, DetectionNetwork, setFromModelzoo))
```
Comment (moratom) [inline bindings/python/src/pipeline/node/DetectionNetworkBindings.cpp:81]: Let's skip this overload, hopefully we'll get the automatic number of shaves in soon and in the mean time people can set the number through the more "advanced API".

File: bindings/python/src/pipeline/node/DetectionParserBindings.cpp [hunk 2]
```diff
@@ -41,6 +41,10 @@ void bind_detectionparser(pybind11::module& m, void* pCallstack){
         .def("getNumFramesPool", &DetectionParser::getNumFramesPool, DOC(dai, node, DetectionParser, getNumFramesPool))
         .def("setBlob", py::overload_cast<dai::OpenVINO::Blob>(&DetectionParser::setBlob), py::arg("blob"), DOC(dai, node, DetectionParser, setBlob))
         .def("setBlob", py::overload_cast<const dai::Path&>(&DetectionParser::setBlob), py::arg("path"), DOC(dai, node, DetectionParser, setBlob, 2))
+        .def("setNNArchive", py::overload_cast<const NNArchive&>(&DetectionParser::setNNArchive), py::arg("nnArchive"), DOC(dai, node, DetectionParser, setNNArchive))
+        .def("setNNArchive", py::overload_cast<const NNArchive&, int>(&DetectionParser::setNNArchive), py::arg("nnArchive"), py::arg("numShaves"), DOC(dai, node, DetectionParser, setNNArchive, 2))
```
Comment (moratom) [inline bindings/python/src/pipeline/node/DetectionParserBindings.cpp:45]: The detection parser doesn't care about the number of shaves.

File: include/depthai/modelzoo/Zoo.hpp [hunk 3]
```diff
@@ -26,4 +26,12 @@ std::string getModelFromZoo(const NNModelDescription& modelDescription,
...
  */
 void downloadModelsFromZoo(const std::string& path, const std::string& cacheDirectory = MODEL_ZOO_DEFAULT_CACHE_DIRECTORY);
+
+/**
+ * @brief Remove cache folder
+ *
+ * @param cacheDirectory: Cache directory to remove, default is MODEL_ZOO_DEFAULT_CACHE_DIRECTORY
+ */
+void removeCacheFolder(const std::string& cacheDirectory = MODEL_ZOO_DEFAULT_CACHE_DIRECTORY);
+
```
Comment (moratom) [inline include/depthai/modelzoo/Zoo.hpp:36]: Who calls this?

**Review by moratom — COMMENTED** [review]
Thanks Jakub!

Let's remove the overloads for setting number of shaves, people can still use the lower level API if they will need it and for the default case we'll hopefully get smart defaults in soon.

File: bindings/python/src/pipeline/node/DetectionParserBindings.cpp (hunk 2 above)
Comment (lnotspotl) [reply bindings/python/src/pipeline/node/DetectionParserBindings.cpp]: Removed

File: bindings/python/src/pipeline/node/DetectionNetworkBindings.cpp (hunk 1 above)
Comment (lnotspotl) [reply bindings/python/src/pipeline/node/DetectionNetworkBindings.cpp]: Removed

File: src/pipeline/node/DetectionNetwork.cpp [hunk 4]
```diff
@@ -22,24 +23,19 @@ namespace node {
...
-class DetectionNetwork::Impl {
-   public:
-    Impl() = default;
-
-    /*
-     * Place for future private stuff.
-     */
+DetectionNetwork::DetectionNetwork(const std::shared_ptr<Device>& device)
+    : DeviceNodeGroup(device), out{detectionParser->out}, outNetwork{neuralNetwork->out}, input{neuralNetwork->input}, passthrough{neuralNetwork->passthrough} {
+    setDeviceForAllDeviceNodes(device);
```
Comment (moratom) [inline src/pipeline/node/DetectionNetwork.cpp:28]: Shouldn't `setDeviceForAllDeviceNodes(device)` be called in the `DeviceNodeGroup` constructor? Although I guess at that stage, the subnodes don't exist yet.

**Pull request merged**